File: android/src/main/java/com/visioncameramlkit/infrastructure/mlkit/adapters/MLKitTextAdapter.kt
package com.visioncameramlkit.infrastructure.mlkit.adapters

import com.google.mlkit.vision.text.Text
import com.visioncameramlkit.domain.models.BoundingBox
import com.visioncameramlkit.domain.models.Corner
import com.visioncameramlkit.domain.models.TextBlock
import com.visioncameramlkit.domain.models.TextElement
import com.visioncameramlkit.domain.models.TextLine
import com.visioncameramlkit.domain.models.TextRecognitionResult
import com.visioncameramlkit.domain.models.TextSymbol

object MLKitTextAdapter {
  fun toDomain(mlkitText: Text?): TextRecognitionResult =
    TextRecognitionResult(
      text = mlkitText?.text,
      blocks = mlkitText?.textBlocks?.map { toTextBlock(it) } ?: emptyList(),
    )

  private fun toTextBlock(block: Text.TextBlock): TextBlock =
    TextBlock(
      text = block.text,
      bounds = block.boundingBox?.let { toBoundingBox(it) },
      corners = block.cornerPoints?.map { toCorner(it) },
      lines = block.lines.map { toTextLine(it) },
      languages = listOf(block.recognizedLanguage),
    )

  private fun toTextLine(line: Text.Line): TextLine =
    TextLine(
      text = line.text,
      bounds = line.boundingBox?.let { toBoundingBox(it) },
      corners = line.cornerPoints?.map { toCorner(it) },
      elements = line.elements.map { toTextElement(it) },
      confidence = line.confidence.toDouble(),
      angle = line.angle.toDouble(),
      languages = listOf(line.recognizedLanguage),
    )

  private fun toTextElement(element: Text.Element): TextElement =
    TextElement(
      text = element.text,
      bounds = element.boundingBox?.let { toBoundingBox(it) },
      corners = element.cornerPoints?.map { toCorner(it) },
      symbols = element.symbols.map { toTextSymbol(it) },
      confidence = element.confidence.toDouble(),
      angle = element.angle.toDouble(),
      languages = listOf(element.recognizedLanguage),
    )

  private fun toTextSymbol(symbol: Text.Symbol): TextSymbol =
    TextSymbol(
      text = symbol.text,
      bounds = symbol.boundingBox?.let { toBoundingBox(it) },
      corners = symbol.cornerPoints?.map { toCorner(it) },
      confidence = symbol.confidence.toDouble(),
      angle = symbol.angle.toDouble(),
      languages = listOf(symbol.recognizedLanguage),
    )

  private fun toBoundingBox(rect: android.graphics.Rect): BoundingBox =
    BoundingBox(
      x = rect.exactCenterX().toDouble(),
      y = rect.exactCenterY().toDouble(),
      centerX = rect.centerX().toDouble(),
      centerY = rect.centerY().toDouble(),
      width = rect.width().toDouble(),
      height = rect.height().toDouble(),
      top = rect.top.toDouble(),
      left = rect.left.toDouble(),
      bottom = rect.bottom.toDouble(),
      right = rect.right.toDouble(),
    )

  private fun toCorner(point: android.graphics.Point): Corner = Corner(x = point.x.toDouble(), y = point.y.toDouble())
}

File: ios/Infrastructure/MLKit/Adapters/MLKitTextAdapter.swift
import Foundation

#if MLKIT_TEXT_RECOGNITION_ANY
  import MLKitTextRecognition

  class MLKitTextAdapter {

    static func toDomain(_ mlkitText: Text?) -> DomainTextRecognitionResult {
      return DomainTextRecognitionResult(
        text: mlkitText?.text,
        blocks: mlkitText?.blocks.map { toTextBlock($0) } ?? []
      )
    }

    private static func toTextBlock(_ block: MLKitTextRecognition.TextBlock)
      -> DomainTextBlock
    {
      return DomainTextBlock(
        text: block.text,
        bounds: toBoundingBox(block.frame),
        corners: block.cornerPoints.map { toCorner($0.cgPointValue) },
        lines: block.lines.map { toTextLine($0) },
        languages: block.recognizedLanguages.map { $0.languageCode ?? "" }
      )
    }

    private static func toTextLine(_ line: MLKitTextRecognition.TextLine)
      -> DomainTextLine
    {
      return DomainTextLine(
        text: line.text,
        bounds: toBoundingBox(line.frame),
        corners: line.cornerPoints.map { toCorner($0.cgPointValue) },
        elements: line.elements.map { toTextElement($0) },
        confidence: nil,
        angle: nil,
        languages: line.recognizedLanguages.map { $0.languageCode ?? "" }
      )
    }

    private static func toTextElement(_ element: MLKitTextRecognition.TextElement)
      -> DomainTextElement
    {
      return DomainTextElement(
        text: element.text,
        bounds: toBoundingBox(element.frame),
        corners: element.cornerPoints.map { toCorner($0.cgPointValue) },
        symbols: [],
        confidence: nil,
        angle: nil,
        languages: element.recognizedLanguages.map { $0.languageCode ?? "" }
      )
    }

    private static func toBoundingBox(_ bounds: CGRect) -> DomainBoundingBox {
      let offsetX = (bounds.midX - ceil(bounds.width)) / 2.0
      let offsetY = (bounds.midY - ceil(bounds.height)) / 2.0

      let x = bounds.maxX + offsetX
      let y = bounds.minY + offsetY

      return DomainBoundingBox(
        x: bounds.midX + (bounds.midX - x),
        y: bounds.midY + (y - bounds.midY),
        centerX: bounds.midX,
        centerY: bounds.midY,
        width: bounds.width,
        height: bounds.height,
        top: bounds.maxY,
        left: bounds.minX,
        bottom: bounds.minY,
        right: bounds.maxX
      )
    }

    private static func toCorner(_ point: CGPoint) -> DomainCorner {
      return DomainCorner(x: Double(point.x), y: Double(point.y))
    }
  }
#endif  // MLKIT_TEXT_RECOGNITION_ANY
